packages/perf/src/index.ts

```diff
@@ -0,0 +1,172 @@
+import { logger } from '@libp2p/logger'
+import { anySignal } from 'any-signal'
+import { MAX_INBOUND_STREAMS, PROTOCOL_NAME, TIMEOUT, WRITE_BLOCK_SIZE } from './constants.js'
+import type { Connection } from '@libp2p/interface/src/connection/index.js'
+import type { Startable } from '@libp2p/interface/startable'
+import type { ConnectionManager } from '@libp2p/interface-internal/connection-manager'
+import type { IncomingStreamData, Registrar } from '@libp2p/interface-internal/registrar'
+import type { AbortOptions } from '@libp2p/interfaces'
+
+const log = logger('libp2p:perf')
+
+export const defaultInit: PerfServiceInit = {
+  protocolName: '/perf/1.0.0',
+  maxInboundStreams: 1 << 10,
+  maxOutboundStreams: 1 << 10,
+  timeout: 10000,
+  writeBlockSize: BigInt(64 << 10)
+}
+
+export interface PerfService {
+  perf: (connection: Connection, sendBytes: bigint, recvBytes: bigint, options?: AbortOptions) => Promise<void>
+  measureDownloadBandwidth: (connection: Connection, size: bigint) => Promise<number>
+  measureUploadBandwidth: (connection: Connection, size: bigint) => Promise<number>
+}
+
+export interface PerfServiceInit {
+  protocolName?: string
+  maxInboundStreams?: number
+  maxOutboundStreams?: number
+  timeout?: number
+  writeBlockSize?: bigint
+}
+
+export interface PerfServiceComponents {
+  registrar: Registrar
+  connectionManager: ConnectionManager
+}
+
+export class DefaultPerfService implements Startable, PerfService {
+  public readonly protocol: string
+  private readonly components: PerfServiceComponents
+  private started: boolean
+  private readonly databuf: ArrayBuffer
+  private readonly maxInboundStreams: number
+  private readonly maxOutboundStreams: number
+  private readonly timeout: number
+  private readonly writeBlockSize: bigint
+
+  constructor (components: PerfServiceComponents, init: PerfServiceInit) {
+    this.components = components
+    this.started = false
+    this.protocol = init.protocolName ?? PROTOCOL_NAME
+    this.databuf = new ArrayBuffer(Number(init.writeBlockSize ?? WRITE_BLOCK_SIZE))
+    this.maxInboundStreams = init.maxInboundStreams ?? MAX_INBOUND_STREAMS
+    this.maxOutboundStreams = init.maxOutboundStreams ?? MAX_INBOUND_STREAMS
+    this.timeout = init.timeout ?? TIMEOUT
```

Comment:
Why do you need a timeout? The caller may set one if they care for it. We don't need one in here.